pallets/otc-settlements/src/lib.rs

```diff
@@ -0,0 +1,542 @@
+use sp_runtime::offchain::storage_lock::{StorageLock, Time};
+use sp_runtime::offchain::Duration;
+use sp_runtime::traits::AccountIdConversion;
+use sp_std::vec;
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarks;
+
+pub mod weights;
+
+use weights::WeightInfo;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+pub type Balance = u128;
+pub type NamedReserveIdentifier = [u8; 8];
+
+pub const PALLET_ID: PalletId = PalletId(*b"otcsettl");
+
+pub const NAMED_RESERVE_ID: NamedReserveIdentifier = *b"otcorder";
+
+// value taken from https://github.com/substrate-developer-hub/recipes/blob/master/pallets/ocw-demo/src/lib.rs
+pub const UNSIGNED_TXS_PRIORITY: u64 = 100;
+
+pub const OFFCHAIN_WORKER_DATA: &[u8] = b"hydradx/otc-settlements/data/";
+pub const OFFCHAIN_WORKER_DATA_LAST_UPDATE: &[u8] = b"hydradx/otc-settlements/data-last-update/";
+pub const OFFCHAIN_WORKER_LOCK: &[u8] = b"hydradx/otc-settlements/lock/";
+pub const OFFCHAIN_WORKER_MAX_ITERATIONS: &[u8] = b"hydradx/otc-settlements/max-iterations/";
+pub const LOCK_DURATION: u64 = 5_000; // 5 seconds
+pub const NUM_OF_ITERATIONS: u32 = 40;
+
+pub type AssetIdOf<T> = <T as pallet_otc::Config>::AssetId;
+
+type SortedOtcsStorageType = (OrderId, FixedU128, FixedU128, FixedU128);
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_otc::Config + SendTransactionTypes<Call<Self>> {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Named reservable multi currency
+		type Currency: MultiCurrency<Self::AccountId, CurrencyId = AssetIdOf<Self>, Balance = Balance>;
+
+		///Router implementation
+		type Router: RouteProvider<AssetIdOf<Self>>
+			+ RouterT<Self::RuntimeOrigin, AssetIdOf<Self>, Balance, Trade<AssetIdOf<Self>>, AmountInAndOut<Balance>>
+			+ RouteSpotPriceProvider<AssetIdOf<Self>>;
+
+		/// Provider of existential deposits.
+		type ExistentialDeposits: GetByKey<AssetIdOf<Self>, Balance>;
+
+		/// Determines the minimum profit
+		#[pallet::constant]
+		type ExistentialDepositMultiplier: Get<u8>;
+
+		/// Account who receives the profit.
+		#[pallet::constant]
+		type ProfitReceiver: Get<Self::AccountId>;
+
+		/// Determines when we consider an arbitrage as closed.
+		#[pallet::constant]
+		type PricePrecision: Get<FixedU128>;
+
+		/// Router weight information.
+		type RouterWeightInfo: AmmTradeWeights<Trade<AssetIdOf<Self>>>;
+
+		/// Weight information for the extrinsics.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn offchain_worker(block_number: BlockNumberFor<T>) {
+			// limit the cases when the offchain worker run
+			if sp_io::offchain::is_validator() {
```

Comment:
why you limit it tho? i assume only validators (collators) can actually settle it with unsigned tx?

Reply:
additional spam prevention, if you run it as a non validator.